// togame/dist/ui/base.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserInterfaceBase = void 0;
const blessed_1 = __importDefault(require("blessed"));
const events_1 = __importDefault(require("events"));
const config_1 = require("../config");
/**
 * 基础画布结构
 *
 * @export
 * @class UserInterfaceBase
 * @extends {EventEmitter}
 * @implements {ISetable}
 * @implements {IDisposable}
 */
class UserInterfaceBase extends events_1.default {
    constructor() {
        super(...arguments);
        this.screen = blessed_1.default.screen({
            smartCSR: true,
            autoPadding: true,
            title: 'Game >_<#@!'
        });
    }
    /**
     * 初始化
     *
     * @param {number} width
     * @param {number} height
     * @memberof UserInterfaceBase
     */
    setup(width, height) {
        this.initEvents();
        this.initLayouts(width, height);
    }
    initEvents() {
        this.screen.key(['escape', 'C-c'], () => {
            return process.exit(0);
        });
        this.screen.on('keypress', (_, key) => {
            this.emit(key.name);
        });
    }
    initLayouts(width, height) {
        this.layout = blessed_1.default.box({
            parent: this.screen,
            top: 'center',
            left: 'center',
            width: width * config_1.config.WIDTH_SCALE,
            height: height,
            tags: true,
            border: {
                type: 'bg'
            },
            style: {
                bg: '#000'
            }
        });
    }
    /**
     * 重绘
     *
     * @memberof UserInterfaceBase
     */
    draw() {
        this.screen.render();
    }
    /**
     * 释放资源
     *
     * @memberof UserInterfaceBase
     */
    dispose() {
        this.screen.destroy();
    }
}
exports.UserInterfaceBase = UserInterfaceBase;
