// togame/dist/ui/GameUI.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GameUI = void 0;
const blessed_1 = __importDefault(require("blessed"));
const config_1 = require("../config");
const base_1 = require("./base");
class GameUI extends base_1.UserInterfaceBase {
    initLayouts(width, height) {
        const widPrompt = 8;
        super.initLayouts(width + widPrompt + 4, height + 4);
        this.playground = blessed_1.default.box({
            parent: this.layout,
            top: 1,
            left: 4,
            width: width * config_1.config.WIDTH_SCALE,
            height: height,
            border: {
                type: 'line'
            },
            style: {
                border: {
                    fg: '#fff',
                    bg: '#000'
                },
                bg: '#000'
            }
        });
        this.tipBox = blessed_1.default.box({
            parent: this.layout,
            top: 1,
            right: 4,
            width: widPrompt * config_1.config.WIDTH_SCALE,
            height: 4,
            align: 'center',
            valign: 'middle',
            content: ['Ctrl: `↑ ↓ ← →`', 'Exit: `ctrl + c`'].join('\n'),
            border: {
                type: 'line'
            },
            style: {
                border: {
                    fg: '#fff',
                    bg: '#000'
                },
                bg: '#000'
            }
        });
        this.promptBox = blessed_1.default.box({
            parent: this.layout,
            top: 5,
            right: 4,
            width: widPrompt * config_1.config.WIDTH_SCALE,
            height: 3,
            align: 'center',
            valign: 'middle',
            border: {
                type: 'line'
            },
            style: {
                border: {
                    fg: '#fff',
                    bg: '#000'
                },
                bg: '#000'
            }
        });
        this.promptBox.hide();
        this.gameoverBox = blessed_1.default.box({
            parent: this.screen,
            top: 'center',
            left: 'center',
            width: 20,
            height: 6,
            tags: true,
            valign: 'middle',
            content: `{center}Game Over!\n\nPress enter to try again{/center}`,
            border: {
                type: 'line'
            },
            style: {
                fg: 'black',
                bg: 'magenta',
                border: {
                    fg: '#ffffff'
                }
            }
        });
        this.hideGameOver();
    }
    setPromptContents(contents = []) {
        if (!contents.length) {
            this.promptBox.hide();
            return;
        }
        this.promptBox.setContent(contents.join('\n'));
        this.promptBox.height = contents.length + 2;
        this.promptBox.show();
    }
    setRects(points) {
        this.playground.children.forEach(child => child.destroy());
        this.playground.children = [];
        points.forEach(p => {
            blessed_1.default.box({
                parent: this.playground,
                top: p.y,
                left: p.x * config_1.config.WIDTH_SCALE,
                width: 1 * config_1.config.WIDTH_SCALE,
                height: 1,
                style: {
                    bg: p.color || '#2ad'
                }
            });
        });
    }
    showGameOver() {
        this.gameoverBox.show();
        this.draw();
    }
    hideGameOver() {
        this.gameoverBox.hide();
        this.draw();
    }
}
exports.GameUI = GameUI;
