// togame/dist/ui/GameListUI.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GameListUI = void 0;
const blessed_1 = __importDefault(require("blessed"));
const common_1 = require("../common");
const config_1 = require("../config");
const base_1 = require("./base");
const TITLE_HEIGHT = 3;
const OPTION_WIDTH = 12;
const OPTION_HEIGHT = 3;
class GameListUI extends base_1.UserInterfaceBase {
    constructor() {
        super(...arguments);
        this.options = [];
        this.focusIndex = 0;
        this.boxes = [];
    }
    choseOptions(options) {
        this.options = options;
        super.setup(OPTION_WIDTH * 2, OPTION_HEIGHT * (options.length + 2) + TITLE_HEIGHT);
        this.layout.border.type = 'line';
        this.layout.style.border.bg = '#000';
        this.createTitleBox();
        this.createArrow();
        this.createOptionBoxes();
        this.bindFocus();
        this.draw();
        return new Promise(resolve => {
            this.on('enter', () => {
                this.dispose();
                resolve(this.options[this.focusIndex]);
            });
        });
    }
    createTitleBox() {
        this.titleBox = blessed_1.default.box({
            parent: this.layout,
            top: 1,
            left: 'center',
            width: '90%',
            height: TITLE_HEIGHT,
            align: 'center',
            valign: 'middle',
            content: `========== It's time to game! ==========`,
            tags: true,
            style: {
                fg: '#000',
                bg: '#fff'
            }
        });
    }
    createArrow() {
        if (!this.arrow) {
            this.arrow = blessed_1.default.box({
                parent: this.layout,
                top: OPTION_HEIGHT * this.focusIndex + OPTION_HEIGHT + TITLE_HEIGHT,
                left: 0,
                width: (OPTION_WIDTH * config_1.config.WIDTH_SCALE) / 2,
                height: OPTION_HEIGHT,
                align: 'right',
                valign: 'middle',
                content: ' ===> ',
                style: {
                    fg: '#fff',
                    bg: '#000'
                }
            });
            return;
        }
        this.arrow.top = OPTION_HEIGHT * this.focusIndex + OPTION_HEIGHT + TITLE_HEIGHT;
    }
    createOptionBoxes() {
        for (let i = 0; i < this.options.length; i++) {
            const box = blessed_1.default.box({
                parent: this.layout,
                top: OPTION_HEIGHT * i + OPTION_HEIGHT + TITLE_HEIGHT,
                left: (OPTION_WIDTH * config_1.config.WIDTH_SCALE) / 2 + 3,
                width: OPTION_WIDTH * config_1.config.WIDTH_SCALE,
                height: OPTION_HEIGHT,
                align: 'center',
                valign: 'middle',
                content: this.options[i].label,
                border: {
                    type: 'line'
                },
                style: {
                    border: {
                        fg: '#666',
                        bg: '#000'
                    },
                    fg: '#000',
                    bg: '#666',
                    focus: {
                        fg: '#000',
                        bg: '#fff',
                        border: {
                            fg: '#fff'
                        }
                    }
                }
            });
            this.boxes.push(box);
        }
    }
    bindFocus() {
        const setFocus = (index) => {
            var _a;
            this.focusIndex = index;
            if (this.focusIndex < 0) {
                this.focusIndex = this.options.length - 1;
            }
            if (this.focusIndex >= this.options.length) {
                this.focusIndex = 0;
            }
            (_a = this.boxes[this.focusIndex]) === null || _a === void 0 ? void 0 : _a.focus();
            this.createArrow();
            this.draw();
        };
        this.on(common_1.EDirection.UP, () => {
            setFocus(this.focusIndex - 1);
        });
        this.on(common_1.EDirection.DOWN, () => {
            setFocus(this.focusIndex + 1);
        });
        setFocus(0);
    }
}
exports.GameListUI = GameListUI;
