// togame/dist/ui/ScaleUI.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScaleUI = void 0;
const blessed_1 = __importDefault(require("blessed"));
const common_1 = require("../common");
const config_1 = require("../config");
const base_1 = require("./base");
class ScaleUI extends base_1.UserInterfaceBase {
    setup() {
        const WID = 12;
        super.setup(WID, WID);
        this.playground = blessed_1.default.box({
            parent: this.layout,
            top: 'center',
            left: 'center',
            width: WID * config_1.config.WIDTH_SCALE,
            height: WID,
            align: 'center',
            valign: 'middle',
            content: [
                //
                'Press `up/down`,',
                'to make this square,',
                '',
                'then `enter` to confirm'
            ].join('\n'),
            border: {
                type: 'bg'
            },
            style: {
                border: {
                    bg: '#2ad'
                },
                fg: '#000',
                bg: '#2ad'
            }
        });
        this.on(common_1.EDirection.UP, () => {
            config_1.config.WIDTH_SCALE++;
            this.playground.width = WID * config_1.config.WIDTH_SCALE;
            this.draw();
        });
        this.on(common_1.EDirection.DOWN, () => {
            config_1.config.WIDTH_SCALE = Math.max(config_1.config.WIDTH_SCALE - 1, 1);
            this.playground.width = WID * config_1.config.WIDTH_SCALE;
            this.draw();
        });
        this.draw();
        return new Promise(resolve => {
            this.on('enter', () => {
                this.dispose();
                resolve();
            });
        });
    }
}
exports.ScaleUI = ScaleUI;
